compiler/src/org.graalvm.compiler.core.aarch64/src/org/graalvm/compiler/core/aarch64/AArch64NodeMatchRules.java

```diff
@@ -233,38 +266,68 @@ public ComplexMatchResult extendedPointerAddShift(AArch64PointerAddNode addP) {
+    public ComplexMatchResult unsignedBitField(BinaryNode shift, ValueNode value, ConstantNode a, ConstantNode b) {
+        JavaKind kind = shift.getStackKind();
+        return emitUnsignedBitField(kind, shift, value, a, b);
+    }
 
-        // The Java(R) Language Specification CHAPTER 15.19 Shift Operators says:
-        // "If the promoted type of the left-hand operand is int(long), then only the five(six)
-        // lowest-order bits of the right-hand operand are used as the shift distance."
-        distance = distance & (srcKind == JavaKind.Int ? 0x1f : 0x3f);
+    @MatchRule("(LeftShift=shift (ZeroExtend=extend (And value Constant=b)) Constant=a)")
+    @MatchRule("(ZeroExtend=extend (And (UnsignedRightShift=shift value Constant=a) Constant=b))")
+    @MatchRule("(ZeroExtend=extend (LeftShift=shift (And value Constant=b) Constant=a))")
+    public ComplexMatchResult unsignedExtBitField(ZeroExtendNode extend, BinaryNode shift, ValueNode value, ConstantNode a, ConstantNode b) {
+        JavaKind kind = extend.getStackKind();
+        return emitUnsignedBitField(kind, shift, value, a, b);
+    }
 
-        // Constraint 1: Mask plus one should be a power-of-2 integer.
-        if (!CodeUtil.isPowerOf2(mask + 1)) {
-            return null;
-        }
-        int width = CodeUtil.log2(mask + 1);
+    @MatchRule("(LeftShift=shift (SignExtend value) Constant)")
+    public ComplexMatchResult signedBitField(LeftShiftNode shift) {
+        JavaKind kind = shift.getStackKind();
+        assert kind.isNumericInteger();
+
+        SignExtendNode extend = (SignExtendNode) shift.getX();
+        int srcBits = extend.getInputBits();
+        int resultBits = extend.getResultBits();
+        assert kind.getBitCount() == resultBits;
+
+        int lsb = shift.getY().asJavaConstant().asInt() & (resultBits - 1);
+        // Get the min value of the srcBits and (resultBits - lsb) as the bitfield width.
+        int width = Math.min(srcBits, resultBits - lsb);
+        assert width >= 1 && width <= resultBits - lsb;
+
+        ValueNode value = extend.getValue();
+        return emitBitField(kind, BitFieldOpCode.SBFIZ, value, lsb, width);
+    }
+
+    @MatchRule("(RightShift=rshift (LeftShift=lshift value Constant) Constant)")
+    @MatchRule("(UnsignedRightShift=rshift (LeftShift=lshift value Constant) Constant)")
+    public ComplexMatchResult bitFieldMove(BinaryNode rshift, LeftShiftNode lshift) {
+        JavaKind srcKind = rshift.getStackKind();
+        assert srcKind.isNumericInteger();
         int srcBits = srcKind.getBitCount();
-        // Constraint 2: Bit field width is less than 31(63) for int(long) as any bit field move
-        // operations can be done by a single shift instruction if the width is 31(63).
-        if (width >= srcBits - 1) {
+        int lshiftNum = lshift.getY().asJavaConstant().asInt() & (srcBits - 1);
+        int rshiftNum = rshift.getY().asJavaConstant().asInt() & (srcBits - 1);
```

Comment:
This may be missing the masking of the lower bits as well. (Same as above in `.signedBitField`) Please doublecheck.

Reply:
Actually I think ` & (srcBits - 1)` is not needed, since the compiler has handled it in the mid-end.